core/adapters/base.ts

```diff
@@ -22,18 +23,56 @@ export abstract class Adapter {
 
   protected insertInto(target: dataform.ITarget, columns: string[], query: string) {
     return `
-      insert into ${this.resolveTarget(target)}
-      (${columns.join(",")})
-      select ${columns.join(",")}
-      from (${query}) as insertions`;
+insert into ${this.resolveTarget(target)}
+(${columns.join(",")})
+select ${columns.join(",")}
+from (${query}) as insertions`;
   }
 
   protected oppositeTableType(type: string) {
     return this.baseTableType(type) === "table" ? "view" : "table";
   }
 
   protected where(query: string, where: string) {
-    return `select * from (${query}) as subquery
-        where ${where || "true"}`;
+    return where
+      ? `
+  select * from (${query}) as subquery
+    where ${where}`
+      : query;
+  }
+
+  protected shouldWriteIncrementally(
+    runConfig: dataform.IRunConfig,
+    tableMetadata?: dataform.ITableMetadata
+  ) {
+    return !runConfig.fullRefresh && tableMetadata && tableMetadata.type !== "view";
+  }
+
+  protected addPreOps(
+    table: dataform.ITable,
+    dataformCoreVersion: string,
+    runConfig: dataform.IRunConfig,
+    tableMetadata: dataform.ITableMetadata
+  ): Task[] {
+    return (dataformCoreVersion > "1.4.8" &&
```

Comment:
IMO we could make this a tad more readable:

```
let preOps = table.preOps;
if (<conditions>) {
  preOps = table.incrementalPreOps;
}
return (preOps || []).map(pre => Task.statement(pre));
```

Reply:
👍